### ISSUE TITLE
google_container_node_pool with name_prefix generates names that are too long

### ISSUE THREAD

**Issue by rvandegrift** [hashicorp/terraform-provider-google#8261]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version
```
Terraform v0.14.4
+ provider registry.terraform.io/hashicorp/google v3.53.0
+ provider registry.terraform.io/hashicorp/google-beta v3.53.0
```

### Affected Resource(s)

* google_container_node_pool

### Terraform Configuration Files

```tf
resource "google_container_node_pool" "nodes" {
  provider           = google-beta
  name_prefix        = "ca-nonprod-core-"
  ...
}
```

### Debug Output

Relevant parts of the debug output are at: https://gist.github.com/rvandegrift/413477c86493a930cfef38b92ffdf65d

### Expected Behavior

Ideally, terraform should truncate the timestamp in the generated suffix.  26 digits of precision seems excessive :)

### Actual Behavior

```
$ terraform apply
...

Error: error creating NodePool: googleapi: Error 400: Node_pool.name must be less than 40 characters., badRequest
```

**Comment by venkykuberan** [comment]
This is more of change in terraform sdk (github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.0/helper/resource/id.go) than the provider. @ScottSuarez  what are your thoughts ?

**Comment by ScottSuarez** [comment]
yeah this is an upstream issue for terraform team. Not much we can do on our side.

[There is an existing request for their team to provide this enhancement.](https://github.com/hashicorp/terraform-plugin-sdk/issues/475)

**Comment by paddycarver** [comment]
Why can't you do anything on your side? [Just don't call the SDK's function.](https://github.com/hashicorp/terraform-provider-google/blob/287a89e3c5086bdb484871b3973a523b5f014672/google/resource_container_node_pool.go#L590)

**Comment by ScottSuarez** [comment]
That's a good point ! I suppose I could add the fix for this but I'm a little hesitant because we wouldn't [gain the mutex locking to ensure the id is unique](https://github.com/hashicorp/terraform-plugin-sdk/blob/651b26a663d34daed50d135ddd25e170ff6c6aae/helper/resource/id.go#L42). I guess it's extremely small edge case that the randomization engine fails so spectacularly. I'll see what my team thinks ! :)

**Comment by ScottSuarez** [comment]
Labeling this as an enhancement

**Comment by paddycarver** [comment]
Using a CSPRNG should make that mostly a non-issue, but there's also no reason you can't use the mutex. Just put a mutex in your global scope and lock and unlock it, which is all the SDK is doing. :D This function doesn't really depend on _anything_ SDK specific.

**Comment by c2thorn** [comment]
A fix for this issue has been submitted in https://github.com/GoogleCloudPlatform/magic-modules/pull/18477, targeting the upcoming **Terraform Google Provider 8.0.0 major release**.

The chosen resolution adapts the same approach used for [compute instance template name prefixes #11448](https://github.com/GoogleCloudPlatform/magic-modules/pull/11448): 
prefixes longer than 14 characters will use a shortened 9-character UUID suffix, extending the maximum allowed `name_prefix` length from 14 to 31 characters for both `google_container_node_pool` and `google_container_cluster.node_pool`.

Because this is a significant change in behavior, it is targeted for the 8.0.0 major release with accompanying upgrade guidance.

**Comment by c2thorn** [comment]
closed with https://github.com/GoogleCloudPlatform/magic-modules/pull/18477

The change should be a part of the 8.0.0 release on 8/25
